### PR TITLE
amazon-cloudwatch-agent: let users specify configuration file paths

### PR DESCRIPTION
**Pull request by commiterate**
Let users specify configuration file paths.

This is to allow dynamic configuration at runtime (mostly during boot since the systemd unit won't restart unless something runs `nixos-rebuild switch`/`test`).

Without this feature, users creating NixOS VM images need to create 1 image per agent configuration permutation if they want immutable images (i.e. don't run `nixos-rebuild switch` on boot).

This can result in a lot of near-identical images whose only difference is a few options in their CloudWatch Agent configuration (e.g. region option for multi-region AWS deployments). Since VM images are usually on the order of GBs, we end up with a lot of wasted space in Nix stores.

With this feature, users can write their own systemd units that dynamically create configurations (e.g. using information from IMDS, AWS SSM Parameter Store, or AWS Secrets Manager) and write them to a static path.

The tradeoff, however, is that mutable configuration files can cause drift if a change isn't followed by a `systemctl restart amazon-cloudwatch-agent.service`. `nixos-rebuild` won't work because we can't access files outside the Nix store in restricted evaluation mode (e.g. can't hash the file and feed it into `restartTriggers`).

Also doing some small cleanup.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

#### Maintainer Pings

@philipmw

### CODE REVIEW

File: nixos/modules/services/monitoring/amazon-cloudwatch-agent.nix [hunk 1]
```diff
@@ -140,38 +193,39 @@ in
...
-        cfg.mode
-      ];
+      restartTriggers =
+        [
+          cfg.package
+          cfg.mode
+          configurationDirectory
+        ]
+        # If the file exists, use hashes instead of contents since contents may be long.
+        ++ (map (file: if (builtins.pathExists file) then (builtins.hashFile "sha256" file) else file) [
```
Comment (arianvp) [inline nixos/modules/services/monitoring/amazon-cloudwatch-agent.nix:225]: I don't understand what you're doing here.

restartTriggers can only be files on the nix store. Which by definition are already a hash .

And there is no point in ever putting a non-existent nix store path in restart triggers.

Comment (arianvp) [reply nixos/modules/services/monitoring/amazon-cloudwatch-agent.nix:212]: IIRC we don't allow `builtins.fileExists` in nixpkgs as it doesn't work in restricted eval mode

**Comment by philipmw** [comment]
This makes sense to me, but I am concerned about someone specifying both a file and an inline config and having the inline config silently ignored. (I do see you document that the file takes precedence.) Do you think we should error out if both are specified?

**Comment by philipmw** [comment]
Sounds good to me; thanks for the thoughtful reply. I approve, though (for some reason) my approval isn't requested.

**Comment by arianvp** [comment]
you can use codeowners instead for modules. it's indeed not supported

**Review by philipmw — APPROVED** [review]

**Review by philipmw — APPROVED** [review]

**Pull request merged**